File: lab1_ciphers/s_vernam.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>

#define PORT 8080
#define BUFFER_SIZE 1024

// Vernam cipher encryption function
void encrypt(char *plaintext, char *key, char *ciphertext)
{
    int len = strlen(plaintext);
    for (int i = 0; i < len; i++)
    {
        ciphertext[i] = ((plaintext[i] ^ key[i]) % 26) + 'A';
    }
    ciphertext[len] = '\0';
}

int main()
{
    int server_fd, new_socket;
    struct sockaddr_in address;
    int opt = 1;
    int addrlen = sizeof(address);

    // Creating socket file descriptor
    if ((server_fd = socket(AF_INET, SOCK_STREAM, 0)) == 0)
    {
        perror("Socket failed");
        exit(EXIT_FAILURE);
    }

    // Forcefully attaching socket to the port 8080
    if (setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt)))
    {
        perror("Setsockopt failed");
        exit(EXIT_FAILURE);
    }

    address.sin_family = AF_INET;
    address.sin_addr.s_addr = INADDR_ANY;
    address.sin_port = htons(PORT);

    // Binding socket
    if (bind(server_fd, (struct sockaddr *)&address, sizeof(address)) < 0)
    {
        perror("Bind failed");
        exit(EXIT_FAILURE);
    }

    // Listening for connections
    if (listen(server_fd, 3) < 0)
    {
        perror("Listen failed");
        exit(EXIT_FAILURE);
    }

    printf("Server is listening on port %d\n", PORT);

    // Accepting connection
    if ((new_socket = accept(server_fd, (struct sockaddr *)&address, (socklen_t *)&addrlen)) < 0)
    {
        perror("Accept failed");
        exit(EXIT_FAILURE);
    }

    // Prepare plaintext and key
    char plaintext[] = "HAI";
    char key[] = "SAY"; // Key must have the same size as plaintext
    char ciphertext[BUFFER_SIZE];

    // Encrypt the plaintext
    encrypt(plaintext, key, ciphertext);
    printf("Plaintext: %s\n", plaintext);
    printf("Key: %s\n", key);
    printf("Ciphertext: %s\n", ciphertext);

    // Send the ciphertext and key to the client
    send(new_socket, ciphertext, strlen(ciphertext), 0);
    send(new_socket, key, strlen(key), 0);

    printf("Encrypted message and key sent to client.\n");

    close(new_socket);
    close(server_fd);
    return 0;
}

File: lab1_ciphers/s_vernam.py
import socket

PORT = 8080
BUFFER_SIZE = 1024

# Vernam cipher encryption function
def encrypt(plaintext, key):
    ciphertext = ""  # Initialize an empty string for ciphertext

    for p, k in zip(plaintext, key):  # Iterate through each character pair in plaintext and key
        p_ascii = ord(p)  # Convert plaintext character to ASCII
        k_ascii = ord(k)  # Convert key character to ASCII

        # XOR operation between plaintext and key character
        xor_result = p_ascii ^ k_ascii

        # Modulo 26 to keep within the English alphabet range
        mod_result = xor_result % 26

        # Convert back to a printable character (A-Z)
        cipher_char = chr(mod_result + ord('A'))

        # Append encrypted character to ciphertext
        ciphertext += cipher_char
    return ciphertext

server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
server_socket.bind(("0.0.0.0", PORT))
server_socket.listen(3)

print(f"Server is listening on port {PORT}")

client_socket, client_address = server_socket.accept()
print(f"Connection established with {client_address}")

# Prepare plaintext and key
plaintext = "HAI"
key = "SAY"

# Encrypt plaintext
ciphertext = encrypt(plaintext, key)
print(f"Plaintext: {plaintext}")
print(f"Key: {key}")
print(f"Ciphertext: {ciphertext}")

# Send ciphertext and key as a single message separated by "::"
message = f"{ciphertext}::{key}"
client_socket.sendall(message.encode())

print("Encrypted message and key sent to client.")

client_socket.close()
server_socket.close()
